sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -618,6 +618,212 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
+      return Stream(this -> false)
+    }
+
+    var ruleApplied = true
+    val (afterRules, autoContinue) = CurrentOrigin.withOrigin(origin) {
+      rule.applyOrElse(this, (_: BaseType) => {
+        ruleApplied = false
+        Stream.empty -> false
+      })
+    }
+    // A stream of a tuple that contains:
+    // - a node that is either the transformed alternative of the current node or the current node,
+    // - a boolean flag if the node was actually transformed,
+    // - a boolean flag if a node's children needs to be transformed to add the node to the valid
+    // alternatives
+    val afterRulesStream = if (afterRules.isEmpty) {
+      if (ruleApplied) {
+        // If the rule returned with empty alternatives then prune
+        Stream.empty
+      } else {
+        // If the rule was not applied then keep the original node
+        Stream((this, false, false))
+      }
+    } else {
+        // If the rule was applied then use the returned alternatives
+        // The alternatives can include the current node and we need to keep track of that
+        var foundEqual = false
+        afterRules.map { afterRule =>
+          (if (this fastEquals afterRule) {
+            foundEqual = true
+            this
+          } else {
+            afterRule.copyTagsFrom(this)
+            afterRule
+          }, true, false)
+        }.append(
+          // If autoContinue is enabled and the current node is not a leaf node and the alternatives
+          // returned by the rule doesn't contain the current node then we need to add the current
+          // node to the stream, but require any of its child nodes to be transformed to keep it as
+          // a valid alternative
+          if (autoContinue && containsChild.nonEmpty && !foundEqual) {
+            Stream((this, false, true))
+          } else {
+            Stream.empty
+          }
+        )
+    }
+
+    def generateChildrenSeq(children: Seq[BaseType]): Stream[(Seq[BaseType], Boolean)] = {
+      children.foldRight(Stream((Seq.empty[BaseType], false)))((child, childrenSeqStream) =>
+        for {
+          (childrenSeq, childrenSeqChanged) <- childrenSeqStream
+          (newChild, childChanged) <- child.multiTransformDownHelper(cond, ruleId)(rule)
+        } yield (newChild +: childrenSeq) -> (childChanged || childrenSeqChanged)
+      )
+    }
+
+    afterRulesStream.flatMap { case (afterRule, transformed, childrenTransformRequired) =>
+      if (afterRule.containsChild.nonEmpty) {
+        generateChildrenSeq(afterRule.children).collect {
+          case (newChildren, childrenTransformed)
+            if !childrenTransformRequired || childrenTransformed =>
+            afterRule.withNewChildren(newChildren) -> (transformed || childrenTransformed)
+        }
+      } else {
+        Stream(afterRule -> transformed)
+      }.map { rewritten_plan =>
+        if (this eq rewritten_plan) {
+          markRuleAsIneffective(ruleId)
```

Comment:
if my understanding is correct, the only way to mark this rule as ineffective is that it returns stream with one alternative which is eq to this ? then why we use `map` here ?

Reply:
Yeah, thanks, this looks wrong. I will fix it with other requests today.

Reply:
Latest commit fixes this, although there is relevant issue that I'm not sure we can mark the rule ineffective if the one element original stream is returned:
https://github.com/apache/spark/pull/38034/files#diff-94575875fbf007fdaf43e4946c69c18649294fed974a46816ab1986f6350541bR691-R699

Reply:
I think it's OK. We can mark the rule as ineffective only if the partial function does not apply.